packages/babel-parser/src/parser/expression.js

```diff
@@ -526,7 +532,23 @@ export default class ExpressionParser extends LValParser {
       }
     }
 
-    return this.parseUpdate(node, update, refExpressionErrors);
+    const expr = this.parseUpdate(node, update, refExpressionErrors);
+
+    const startsExpr = this.hasPlugin("v8intrinsic")
+      ? this.state.type.startsExpr
+      : this.state.type.startsExpr && !this.match(tt.modulo);
+    if (
+      isAwait &&
+      startsExpr &&
+      !this.isAmbiguousAwait() &&
+      (this.scope.inFunction ||
+        (this.hasPlugin("topLevelAwait") && !this.inModule))
+    ) {
+      this.raise(startPos, Errors.AwaitNotInAsyncContext);
```

Comment:
We throw `AwaitNotInAsyncContext` only when `topLevelAwait` is enabled, otherwise a different error message will be provided.

https://github.com/babel/babel/blob/c8117b6a490607cf6aabbdcb8701457211e60e00/packages/babel-parser/src/parser/expression.js#L2328-L2334